### DISCUSSION TITLE
How can I code firmware to use data from motion caputre without kalman estimator?

### DISCUSSION THREAD

**Discussion by cc299792458** [bitcraze/discussions#661] [Q&A] [answered]
Hi,

This Q&A is a succesion of https://github.com/orgs/bitcraze/discussions/654 （Position value returns nan after using ext_pos() to send position data getting from motion capture #654 )

I wonder if I can modify the firmware to use data from motion capture directly without kalman filter. I guess it maybe doesn't need too much code because what I wish is just recieve the data, and set it as state. I wonder where is a suitable place( I mean in which .c file?) to code?
I still want to use ext_pose api to push the data to crazyflie.

Any advice would be appreciate!

**Comment by krichardsson** [comment, marked as answer]
The complementary estimator unfortunately does not have any position information, so you will have to add it.

To roughly outline what you have to do:
1.  Add some variables somewhere where you can store the position you send in with ext_pose.
2. When data arrives on the ext_pose API, store it in the variables.
3. When reading the state, copy the position from the variables to the state.

There are a few ways of doing this, depending on how hackish solution you can accept.

I will outline the easiest solution I can think of, but it will be pretty temporary. I also assume you are only interested in the position (not full pose) but it should be easy to extend if you want to.
1. Add a `static positionMeasurement_t myPosition` to the [stabilizer.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/stabilizer.c) file
2. Also add a function, lets call it `stabilizerStorePosition(const positionMeasurement_t* pos)` that copies the values of `pos` to `myPosition`
3. [In the stabilizer loop, after getting the state](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/stabilizer.c#L284), add some code that copies the position in `myPosition` to the state.
4. Finally you need to call `stabilizerStorePosition()` from the [`extPositionHandler()`](https://github.com/bitcraze/crazyflie-firmware/blob/e6ca813ae493b78a66a1e9ffa5585cbc051f0181/src/modules/src/crtp_localization_service.c#L180) to write the position data to `myPosition`

**Reply by cc299792458** [reply]
Thanks, sir!
